File: LeetCode/HeapPriorityQueue/JavaScript/findMedianFromDataStream.js
// The median is the middle value in an ordered integer list. If the size of the list is even, there is no middle value, and the median is the mean of the two middle values.

// For example, for arr = [2,3,4], the median is 3.
// For example, for arr = [2,3], the median is (2 + 3) / 2 = 2.5.
// Implement the MedianFinder class:

// MedianFinder() initializes the MedianFinder object.
// void addNum(int num) adds the integer num from the data stream to the data structure.
// double findMedian() returns the median of all elements so far. Answers within 10-5 of the actual answer will be accepted.

// Example 1:

// Input
// ["MedianFinder", "addNum", "addNum", "findMedian", "addNum", "findMedian"]
// [[], [1], [2], [], [3], []]
// Output
// [null, null, null, 1.5, null, 2.0]

// Explanation
// MedianFinder medianFinder = new MedianFinder();
// medianFinder.addNum(1);    // arr = [1]
// medianFinder.addNum(2);    // arr = [1, 2]
// medianFinder.findMedian(); // return 1.5 (i.e., (1 + 2) / 2)
// medianFinder.addNum(3);    // arr[1, 2, 3]
// medianFinder.findMedian(); // return 2.0

// Constraints:

// -105 <= num <= 105
// There will be at least one element in the data structure before calling findMedian.
// At most 5 * 104 calls will be made to addNum and findMedian.

// Follow up:

// If all integer numbers from the stream are in the range [0, 100], how would you optimize your solution?
// If 99% of all integer numbers from the stream are in the range [0, 100], how would you optimize your solution?

// TC:
// addNum: O(logn) due to heaps
// findMedian: O(1) only accessing the index 0 of the heap array
// SC:
// addNum: O(n), we have 2 heaps and the total number of elements of both heaps will total all of the elements added
// findMedian: O(1), in place processing

var MedianFinder = function () {
  // store num at each call of addNum
  // Using heaps
  this.maxHeap = new MaxPriorityQueue();
  this.minHeap = new MinPriorityQueue();
};

/**
 * @param {number} num
 * @return {void}
 */
MedianFinder.prototype.addNum = function (num) {
  // add num to internal storage and make sure it's sorted
  // Using min heap and max heap
  // first, add the num to the min queue
  this.minHeap.enqueue(num);

  // second, dequeue the min from minHeap and enqueue to maxHeap
  let minVal = this.minHeap.dequeue();
  this.maxHeap.enqueue(minVal);

  // Let's check if the total number of element we have on both queues is even.
  // It is even if the maxHeap has more element than the minHeap
  // We need to move the top of maxHeap to the minHeap
  if (this.minHeap.size() < this.maxHeap.size()) {
    let maxVal = this.maxHeap.dequeue();
    this.minHeap.enqueue(maxVal);
  }

  // At this point, our two heaps are balanced and should contain the same size of elements
  // console.log( this.minHeap.toArray(), this.maxHeap.toArray() )
};

/**
 * @return {number}
 */
MedianFinder.prototype.findMedian = function () {
  // We have 2 heaps with elements on it
  // If we have an odd number of elements, the size of the minHeap will be greater than the maxHeap,
  // So we just return the top of the minHeap. This will be our median
  if (this.minHeap.size() > this.maxHeap.size()) {
    return this.minHeap.front();
  } else {
    // the total elements we have is even because the maxHeap has more than the minHeap
    // We calculate the mean
    let mean = (this.minHeap.front() + this.maxHeap.front()) / 2;
    return mean;
  }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * var obj = new MedianFinder()
 * obj.addNum(num)
 * var param_2 = obj.findMedian()
 */

// // Example input
// ["MedianFinder","addNum","addNum","addNum","addNum","addNum","addNum","addNum","findMedian","addNum","findMedian"]
// [[],[90203],[78],[234],[23432],[-12123],[8],[5],[],[3],[]]

// This is how the heaps would look like:
// [ 90203 ] [] - add to minheap, then move top of min heap to max heap, size, minHeap size < maxHeap size, move top of maxHeap back to minHeap
// [ 90203 ] [ 78 ] - add to minheap, then move top of min heap to max heap, size, minHeap size !< maxHeap size, do nothing
// [ 234, 90203 ] [ 78 ] - add to minheap, then move top of min heap to max heap, size, minHeap size < maxHeap size, move top of maxHeap back to minHeap
// [ 23432, 90203 ] [ 234, 78 ] - add to minheap, then move top of min heap to max heap, size, minHeap size !< maxHeap size, do nothing
// [ 234, 23432, 90203 ] [ 78, -12123 ] - add to minheap, then move top of min heap to max heap, size, minHeap size < maxHeap size, move top of maxHeap back to minHeap
// [ 234, 23432, 90203 ] [ 78, 8, -12123 ] - add to minheap, then move top of min heap to max heap, size, minHeap size !< maxHeap size, do nothing
// [ 78, 234, 23432, 90203 ] [ 8, 5, -12123 ] - add to minheap, then move top of min heap to max heap, size, minHeap size < maxHeap size, move top of maxHeap back to minHeap
// [ 78, 234, 23432, 90203 ] [ 8, 5, 3, -12123 ] - add to minheap, then move top of min heap to max heap, size, minHeap size !< maxHeap size, do nothing

// This is the expected output
// [null,null,null,null,null,null,null,null,78.00000,null,43.00000]

File: LeetCode/HeapPriorityQueue/python/findMedianFromDataStream.py
from dataclasses import dataclass
from enum import Enum
import heapq
from typing import List


class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"

@dataclass
class medianFinderRecord:
    operation: List[str]
    value: List[List[int]]
    expected: List[str]

class MedianFinder:

    def __init__(self):
        self.my_heapq = heapq # priority queue library
        self.heap_max: List[int] = [] # will be used to store the max priority queue
        self.heap_min: List[int] = [] # will be used to store the min priority queue

    def addNum(self, num: int) -> None:
        self.my_heapq.heappush(self.heap_min, num)

        min_val: int = self.my_heapq.heappop(self.heap_min)
        self.my_heapq.heappush_max(self.heap_max, min_val)

        if len(self.heap_min) < len(self.heap_max):
            maxVal: int = self.my_heapq.heappop_max(self.heap_max)
            self.my_heapq.heappush(self.heap_min, maxVal)

    def findMedian(self) -> float:
        if len(self.heap_min) > len(self.heap_max):
            return self.heap_min[0]
        else:
            mean: float = (self.heap_min[0] + self.heap_max[0]) / 2.0
            return mean

    @staticmethod
    def testSolution(record: medianFinderRecord) -> None:
        print(f"input:\toperations: {record.operation}")
        print(f"\tvalues: {record.value}")
        print(f"expected: {record.expected}")

        obj: MedianFinder = None
        res: List[str] = []

        for i, operation in enumerate(record.operation):
            if operation == "MedianFinder":
                obj = MedianFinder()
                res.append("null")
            elif operation == "addNum":
                obj.addNum(record.value[i][0])
                res.append("null")
            elif operation == "findMedian":
                tmp: float = obj.findMedian()
                formattedVal = f"{tmp:.5f}"
                res.append(formattedVal)
        
        print(f"result: {res}")
        print(Result.PASS.value if res == record.expected else Result.FAIL.value)

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

if __name__ == "__main__":
    records: List[medianFinderRecord] = [
        medianFinderRecord(
            ["MedianFinder","addNum","addNum","findMedian","addNum","findMedian"],
            [[],[1],[2],[],[3],[]],
            ["null","null","null","1.50000","null","2.00000"]
        )
    ]

    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        MedianFinder.testSolution(record)
        print("-" * 50)